packages/salesforcedx-vscode-core/src/sfdxProject/sfdxProjectConfig.ts

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) 2019, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import { SfdxProject, SfdxProjectJson } from '@salesforce/core';
+import { AnyJson, JsonMap } from '@salesforce/ts-types';
+import * as path from 'path';
+import * as vscode from 'vscode';
+
+import { SFDX_PROJECT_FILE } from '../constants';
+import { nls } from '../messages';
+import { notificationService } from '../notifications';
+import { isSfdxProjectOpened } from '../predicates';
+import { SfdxProjectPath } from '../sfdxProject';
+import { telemetryService } from '../telemetry';
+
+/**
+ * Class representing the local sfdx-project.json file.
+ * Does not contain global values.
+ */
+export default class SfdxProjectConfig {
+  private static instance: SfdxProjectJson;
+  private constructor() {
+    throw new Error(
+      'Error: *** call SfdxProject.getInstance() to get the singleton instance of this class ***'
+    );
+  }
+
+  private static async initializeSfdxProjectConfig() {
+    if (
+      !SfdxProjectConfig.instance &&
+      isSfdxProjectOpened.apply(vscode.workspace).result
+    ) {
+      const sfdxProjectPath = SfdxProjectPath.getPath();
+      try {
+        const sfdxProject = await SfdxProject.resolve(sfdxProjectPath);
+        SfdxProjectConfig.instance = await sfdxProject.retrieveSfdxProjectJson();
+        const fileWatcher = vscode.workspace.createFileSystemWatcher(
+          path.join(sfdxProjectPath, SFDX_PROJECT_FILE)
+        );
+        fileWatcher.onDidChange(async () => {
+          try {
+            await SfdxProjectConfig.instance.read();
+          } catch (error) {
+            SfdxProjectConfig.handleError(error);
+            throw error;
+          }
+        });
+      } catch (error) {
+        SfdxProjectConfig.handleError(error);
+        throw error;
+      }
+    }
+  }
+
+  private static handleError(error: any) {
+    let errorMessage = error.message;
+    if (error.name === 'JsonParseError') {
+      errorMessage = nls.localize(
+        'error_parsing_sfdx_project_file',
+        error.path,
+        error.message
+      );
+    }
+    notificationService.showErrorMessage(errorMessage);
+    telemetryService.sendError(errorMessage);
+  }
+
+  public static async getInstance(): Promise<SfdxProjectJson> {
+    if (!SfdxProjectConfig.instance) {
+      await SfdxProjectConfig.initializeSfdxProjectConfig();
+    }
+    return SfdxProjectConfig.instance;
+  }
+
+  public static async getValue(key: string): Promise<AnyJson | undefined> {
```

Comment:
Overall the changes here look good. Since `sfdxProjectConfig.ts` will become the way we interact with `sfdx-project.json` what do we think about accessing its values by doing `SfdxProjectConfig.getValue('packageDirectories')` versus having getters specific to the values the file can have.

I'm asking right now cause we're already using two of those values; packageDirectories and sfdcLoginUrl.

Reply:
@lcampos Are you suggesting that for any value that we're retrieving from the `sfdx-project.json`, we create a specific getter method in this file? For example, since we are already using packageDirectories and sfdcLoginUrl, then we have a method called `SfdxProjectConfig.getPackageDirectories()` and `SfdxProjectConfig.getSfdcLoginUrl()`, and whenever we need to access a new value, we just add another method for that value?

Reply:
I'm not suggesting anything yet, I have some questions regarding how we want/need to consume `sfdx-project.json` data. The reason why I bring this up is because some of these values need to be validated before we can use them in other parts of the project. One example here is how we have some checks in `sfdxPackageDirectories.getPackageDirectoryPaths()`. If I want to consume package directory 
data I would use sfdxPackageDirectories instead of directly calling for that array using `SfdxProjectConfig.getValue('packageDirectories')` because it gives me all the error handling.

Opposite to this, if I need the `sfdcLoginUrl` data, its validations would have to be done in `forceAuthWebLogin` since we do not have something like `sfdxPackageDirectories` for it.

Is this because we need the flexibility of calling `SfdxProjectConfig.getValue('<value name>')` across the project or should we make additional changes so that we use interfaces that will run those validations every time we need to use one of these values ?

Reply:
Update: Alli and I synced "offline", we decided to keep the current changes and monitor how we   access `sfdx-project.json` values before making any further changes. If we need to refactor to make it simpler to consume those values then we'll work on them as needed.